### ISSUE TITLE
Single sync fire event

### ISSUE THREAD

**Issue by plokko** [fico7489/LaravelPivot#19]
Is possible to fire only one event for the sync event instead of multiple attaching/detaching events?
I need to execute an one-time only update after ther relation is updated.

if i use for example
```php 
$model->relation()->attach([1,2,3]);
```
the attached method will be called just one time with `[1,2,3]` ids

```php 
$model->relation()->detach([4,5,6]);
```
the detached method will be called just one time with `[4.5,6]` ids;
but if i use
```php
$model->relation()->sync([1,2,3]);//let's assume the relation was 4,5,6: it will attach 1,2,3 detach 4,5,6
```
pivotDetached will be called 1 time witn `[4,5,6]`, pivotAttached will be called 3 times with `[1]`,`[2]`,`[3]` ids. 

The event should be something like this:
```php
static::pivotModified(function ($model, $relationName, array $pivotIds, $pivotIdsAttributes) {
     // Where pivotids is an array of [ 'attached'=>[ids], 'detached'=>[ids], 'updated'=>[ids]]
}
```
and should be called one-time only for every pivot event ( pivotAttached, pivotDetached, pivotUpdated ) and not only for sync operations.

**Comment by fico7489** [comment]
I have events for attach(), detach() and update(), sync() just calls those 3 methods. Problem is that detach() is called only once for all items, attach() is called in foreach loop for each item : 
https://github.com/laravel/framework/blob/5.2/src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php#L893

Possible solution for you is to aggregate events per request and use them aggregated at the end of the request.

Laravel have 'updating' and 'updated' events for each model, it does not have  'modified' event where 'updated' would be aggregated, but of course you can do it manually...

**Comment by plokko** [comment]
Yes, that's exactly what i want to do: aggregate all the events in just one call but i need to implement it on an event not on the final code because i'm creating a package.

The only way i find to do so is extending the relation how you did on your package, the only think i'm missing is an event fired after a attach/detach/sync operation is executed (and maybe before like pivotModifying and pivotModified, feel free to find a better name).

To be clear:
```php
//fire pivotModifying
$model->relation()->attach([1,2,3]);//fire pivotAttached
//fire pivotModified

//fire pivotModifying
$model->relation()->detach([4,5,6]);//fire pivotDetached
//fire pivotModified

//fire pivotModifying
$model->relation()->sync([1,2,3]);//fire pivotAttached/pivotDetached
//fire pivotModified
```
Just an example draft:
https://github.com/plokko/laravel-pivot/commit/efb12348cf35c75d9f241de94bc67ec23e301637

**Comment by fico7489** [comment]
I understand what you need, but I am not sure if this event is "too much" for package, because all changes on pivot table are already covered with events, your issue is just about aggregating events... I can't say more because I don't know what are you try to achieve in your package...

**Comment by plokko** [comment]
I need to fire a sync event after a pivot modification (with Firebase to be specific), it's not a lightweight operation so repeated calls should be avoided.
If you think it's too much stress for the package i could evaluate an opt-in/opt-out customization throught configuration or i must delegate everything to my package.

**Comment by fico7489** [comment]
maybe you can aggreagate events like this : 

```
public static function boot()
{
    parent::boot();
    
    static::pivotAttached(function ($model, $relationName, $pivotIds, $pivotIdsAttributes) {
        self::fillEvents($model);
    });
    
    static::pivotDetached(function ($model, $relationName, $pivotIds) {
        self::fillEvents($model);
    });
    
    static::pivotUpdated(function ($model, $relationName, $pivotIds, $pivotIdsAttributes) {
        self::fillEvents($model);
    });
}
```

I think that my package is solution for your problem : 
https://github.com/fico7489/laravel-updated-related

You can use some ideas from that package...
